### PR TITLE
minor refactor over EP

### PR DESCRIPTION
**Pull request by tianyu-l**
This PR:

- let `ExpertParallel` handles indices permute / unpermute when EP is used
- move `to_local` to model code to be more explicit
- rename the `expert_parallel` wrapper which does permute / unpermute to `indices_permutation_wrapper` to be more accurate

### CODE REVIEW

File: torchtitan/distributed/expert_parallel.py [hunk 1]
```diff
@@ -216,25 +220,44 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
...
-    tokens each expert gets is a multiple of ALIGN_SIZE_M. The generate_permute_indices
-    kernel also helps achieve this via padding, without incurring synchronization
-    between device and host. Note that this will create side effects when wrapping
-    the for-loop implementation of GroupedExperts, as it does not need padding.
-
-    Among the above:
-    1 and 2 are needed only when expert_parallel_degree > 1.
-    3 is needed even for single-device computation.
-    2 can be moved to ExpertParallel _token_dispatch if not coupled with 3.
+    In order to use torch._grouped_mm, we need to make sure the number of
```
Comment (wwwjn) [inline torchtitan/distributed/expert_parallel.py:257]: nit: This description only talks about padding, didn't talk about generate_permute_indices kernel to permute the inputs to be ordered  by local experts

File: torchtitan/distributed/expert_parallel.py
Comment (wwwjn) [inline torchtitan/distributed/expert_parallel.py:82]: I have question about when we apply _permute() and _unpermute(), now they are applied in 2 places

1. In ExpertParallel(), which is applied on `transformer_block.moe.experts`, so the input of MoE module will be reordered by local experts.

2. When use_grouped_mm is enabled, in `indices_permutation_wrapper`, it will also try to permute the inputs of GroupedExperts by the order of local experts

Why do we need to apply is twice?

File: torchtitan/distributed/expert_parallel.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/distributed/expert_parallel.py:257]: This wrapper now is only responsible for padding, when EP is not used. I renamed to make it more clear.

File: torchtitan/distributed/expert_parallel.py
Comment (tianyu-l) [reply torchtitan/distributed/expert_parallel.py:82]: They won't be applied together.

When EP is used, EP will do `_permute` and `_unpermute`.
When EP is not used, `indices_padding_wrapper` will do them.
